api/include/opentelemetry/trace/trace_state.h

```diff
@@ -45,51 +45,149 @@ class TraceState
+
+    // Set the value for this entry. This overrides the previous value.
+    void SetValue(nostd::string_view value) { value_ = StringToPointer(value); }
 
   private:
     // Store key and value as raw char pointers to avoid using std::string.
     nostd::unique_ptr<const char[]> key_;
     nostd::unique_ptr<const char[]> value_;
+
+    // Copy string into a buffer and return a unique_ptr to the buffer.
+    // This is a workaround for the fact that strcpy doesn't accept a const char* destination.
+    nostd::unique_ptr<const char[]> StringToPointer(nostd::string_view str)
+    {
+      nostd::unique_ptr<char[]> temp(new char[str.size() + 1]);
+      strcpy(temp.get(), str.data());
+      return nostd::unique_ptr<const char[]>(temp.release());
+    }
   };
 
   // An empty TraceState.
-  TraceState() noexcept : num_entries_(0) {}
+  TraceState() noexcept : entries_(new Entry[kMaxKeyValuePairs]), num_entries_(0) {}
 
-  // Returns false if no such key, otherwise returns true and populates value.
-  bool Get(nostd::string_view key, nostd::string_view value) const noexcept { return false; }
+  // Returns false if no such key, otherwise returns true and populates the value parameter with the
+  // associated value.
+  bool Get(nostd::string_view key, nostd::string_view &value) noexcept
+  {
+    for (int i = 0; i < num_entries_; i++)
+    {
+      if (key == (entries_.get())[i].GetKey())
+      {
+        value = (entries_.get())[i].GetValue();
+        return true;
+      }
+    }
+    return false;
+  }
 
   // Creates an Entry for the key-value pair and adds it to entries.
   // If value is null, this function is a no-op.
-  void Set(nostd::string_view key, nostd::string_view value) const noexcept;
+  void Set(nostd::string_view key, nostd::string_view value) noexcept
+  {
+    if (value.data() == NULL)
+      return;
+
+    Entry entry(key, value);
+
+    if (num_entries_ < kMaxKeyValuePairs)
+    {
+      (entries_.get())[num_entries_] = entry;
+      num_entries_++;
+    }
+  }
 
   // Returns true if there are no keys, false otherwise.
-  bool Empty() const noexcept { return true; }
+  bool Empty() const noexcept { return num_entries_ == 0; }
 
   // Returns a span of all the entries. The TraceState object must outlive the span.
-  nostd::span<Entry> Entries() const noexcept { return {}; }
+  nostd::span<Entry> Entries() noexcept { return nostd::span<Entry>(entries_.get(), num_entries_); }
 
   // Returns whether key is a valid key. See https://www.w3.org/TR/trace-context/#key
-  static bool IsValidKey(nostd::string_view key);
+  static bool IsValidKey(nostd::string_view key)
+  {
+    if (key.empty() || key.size() > kKeyMaxSize || !IsLowerCaseAlphaOrDigit(key[0]))
+    {
+      return false;
+    }
+
+    int ats     = 0;
+    const int n = key.size();
+
+    for (int i = 0; i < n; ++i)
+    {
+      char c = key[i];
+      if (!IsLowerCaseAlphaOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/')
+      {
```

Comment:
I think this loop can be just

for(const char c : key) {
...

since i is not used anywhere inside the loop. This would also allow to get rid of n above